File: seminar-07/A/java/A.java
import java.io.*;
import java.util.*;

public class A {
    static class MaxHeap {
        private final ArrayList<Integer> a = new ArrayList<>();

        public int size() {
            return a.size();
        }

        public boolean isEmpty() {
            return a.isEmpty();
        }

        public Integer peek() {
            if (a.isEmpty()) return null;
            return a.get(0);
        }

        public void push(int x) {
            a.add(x);
            shiftUp(a.size() - 1);
        }

        public Integer pop() {
            if (a.isEmpty()) return null;
            int top = a.get(0);
            int last = a.remove(a.size() - 1);
            if (!a.isEmpty()) {
                a.set(0, last);
                shiftDown(0);
            }
            return top;
        }

        private void shiftUp(int i) {
            while (i > 0) {
                int p = (i - 1) / 2;
                if (a.get(p) >= a.get(i)) break;
                int tmp = a.get(p);
                a.set(p, a.get(i));
                a.set(i, tmp);
                i = p;
            }
        }

        private void shiftDown(int i) {
            int n = a.size();
            while (true) {
                int l = 2 * i + 1;
                if (l >= n) break;
                int r = l + 1;
                int j = l;
                if (r < n && a.get(r) > a.get(l)) j = r;
                if (a.get(i) >= a.get(j)) break;
                int tmp = a.get(i);
                a.set(i, a.get(j));
                a.set(j, tmp);
                i = j;
            }
        }
    }

    public static void main(String[] args) throws Exception {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(in.readLine().trim());
        MaxHeap heap = new MaxHeap();
        StringBuilder out = new StringBuilder();

        for (int i = 0; i < n; i++) {
            String[] parts = in.readLine().split(" ");
            if (parts[0].equals("0")) {
                heap.push(Integer.parseInt(parts[1]));
            } else { // "1"
                out.append(heap.pop()).append('\n');
            }
        }

        System.out.print(out.toString());
    }
}

File: seminar-07/A/python/A.py
class MaxHeap:
    def __init__(self, data=None):
        self.a = []
        if data is not None:
            self.a = list(data)
            self.heapify()

    def __len__(self):
        return len(self.a)

    def is_empty(self):
        return len(self.a) == 0

    def peek(self):
        if not self.a:
            return None
        return self.a[0]

    def push(self, x):
        self.a.append(x)
        self.shift_up(len(self.a) - 1)

    def pop(self):
        if not self.a:
            return None
        top = self.a[0]
        last = self.a.pop()
        if self.a:
            self.a[0] = last
            self.shift_down(0)
        return top

    def shift_up(self, i):
        a = self.a
        while i > 0:
            p = (i - 1) // 2
            if a[p] >= a[i]:
                break
            a[p], a[i] = a[i], a[p]
            i = p

    def shift_down(self, i, n=None):
        a = self.a
        if n is None:
            n = len(a)
        while True:
            l = 2 * i + 1
            if l >= n:
                break
            r = l + 1
            j = l
            if r < n and a[r] > a[l]:
                j = r
            if a[i] >= a[j]:
                break
            a[i], a[j] = a[j], a[i]
            i = j

    def heapify(self):
        n = len(self.a)
        for i in range(n // 2 - 1, -1, -1):
            self.shift_down(i, n)

n = int(input())
heap = MaxHeap()
ans = []
for i in range(n):
    line = input().split()
    if line[0] == '0':
        heap.push(int(line[1]))
    elif line[0] == '1':
        ans.append(heap.pop())
print(*ans, sep='\n')